test/divergence_check_test_utils.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include "divergence_block_chain.h"
+
+using namespace celerity;
+using namespace celerity::detail;
+using namespace celerity::detail::divergence_checker_detail;
+
+namespace celerity::detail::divergence_checker_detail {
+struct divergence_block_chain_testspy {
+	static per_node_task_hashes pre_check(divergence_block_chain& div_test, const int max_size) {
+		div_test.add_new_hashes();
+		div_test.collect_hash_counts();
+		return div_test.collect_hashes(max_size);
+	}
+
+	static void post_check(divergence_block_chain& div_test, const int min_size) { div_test.clear(min_size); }
+
+	static void call_check_for_divergence_with_pre_post(std::vector<std::unique_ptr<divergence_block_chain>>& div_test) {
+		std::vector<size_t> sizes;
+		std::transform(div_test.begin(), div_test.end(), std::back_inserter(sizes), [](auto& div) { return div->m_task_records.size(); });
+		auto [min, max] = std::minmax_element(sizes.begin(), sizes.end());
+
+		std::vector<per_node_task_hashes> extended_lifetime_hashes;
```

Comment:
I don't understand what this is for..?

Reply:
The `extended_lifetime_hashes` are needed, because of how the single node `divergence_test_communicator` works. The idea is following:
* each _simulated node_ calls allgather sequentially. The send/receive buffer is saved internally as simple pointers.
* When all _simulated nodes_ called the allgather the data is copied. 

To prevent those send/receive buffers for `per_node_task_hashes` from going out of scope we need to "extend" their lifetime. For this, I included the `extended_lifetime_hashes` vector.